### ISSUE TITLE
Incompatibility between urllib3 and TLS

### ISSUE THREAD

**Issue by fasaxc** [projectcalico/felix#933]
I tested various versions of urllib3 with etcd TLS and OpenStack.  Unfortunately, there were some incompatibilities:
- urllib3 1.7.1, which ships with Ubuntu 14.04 LTS, seemed to cause eventlet hangs.  I had pyOpenSsl installed but I didn't do anything to activate it; according to @Lukasa there's some known issues between pyOpenSsl and eventlet.  It appears that neutron activates pyOpenSsl anyway because I see the memory view error below in the neutron logs (when I run with 1.12).
- No versions of urllib3 support IP-based subjectAltNames, which I tried to use in my certificate.
- urllib3 1.12, which was the PyPI-available version at the time of writing, caused a crash in pyOpenSsl:

```
...
2015-12-03 15:34:28.210 28300 TRACE etcd.client   File "/usr/local/lib/python2.7/dist-packages/urllib3/contrib/pyopenssl.py", line 216, in sendall
2015-12-03 15:34:28.210 28300 TRACE etcd.client     data = memoryview(data)
2015-12-03 15:34:28.210 28300 TRACE etcd.client TypeError: cannot make memory view because object does not have the buffer interface
```

According to @Lukasa, that was introduced in 1.11; it's already fixed in master.

I tested urllib3 master (soon to be 1.13) and that seemed to work without any issues in neutron.

**Comment by fasaxc** [comment]
Would be resolved by move to etcd v3 API #952.

**Comment by neiljerram** [comment]
@fasaxc Is there still any issue here?  I wasn't aware of this, so I wonder if it's a live problem.

**Comment by fasaxc** [comment]
It's fixed in urllib3 1.13 but we don't have a good answer for older versions (such as the one that ships with Ubuntu 14.04).

**Comment by fasaxc** [comment]
I should be able to resolve this as part of the repackaging work that I'm doing.  Also, all the relevant code might be moved to golang by the end of my work there.

**Comment by fasaxc** [comment]
Fixed under #1122
